CRM/Mailing/Event/BAO/MailingEventQueue.php

```diff
@@ -286,23 +291,28 @@ public static function getContactInfo($queueID) {
    * @param null $now
    */
   public static function bulkCreate($params, $now = NULL) {
-    if (!$now) {
-      $now = time();
-    }
+    $cleanNull = function ($value) {
+      return $value === 'null' ? NULL : $value;
+    };
 
-    // construct a bulk insert statement
-    $values = [];
+    // Perform batch-inserts, with batch-sizes limited by BULK_INSERT_COUNT.
+    $rows = [];
     foreach ($params as $param) {
-      $values[] = "( {$param[0]}, {$param[1]}, {$param[2]}, {$param[3]}, '" . substr(sha1("{$param[0]}:{$param[1]}:{$param[2]}:{$param[3]}:{$now}"),
-          0, 16
-        ) . "' )";
+      $rows[] = [
+        'job_id' => $cleanNull($param[0]),
+        'email_id' => $cleanNull($param[1]),
+        'contact_id' => $cleanNull($param[2]),
+        'phone_id' => $cleanNull($param[3]),
+        'hash' => static::hash(),
```

Comment:
you know - we could just have mysql generate the hash here? It might be more efficient - although who knows

Reply:
https://fedingo.com/how-to-generate-random-string-in-mysql/